File: Data Structures & Algorithms/house-robber/submission-4.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        vector<int> max_cost(n+1);
        if (n == 0) return 0;
        if (n == 1) return nums[0];
        max_cost[0] = nums[0];
        max_cost[1] = max(nums[0],nums[1]);
        for(int i = 2; i < n; i++){
            max_cost[i] = max(max_cost[i-1], max_cost[i-2]+nums[i]);
        }
        return max_cost[n-1];
    }
};

File: Data Structures & Algorithms/maximum-product-subarray/submission-4.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        curmin = 1
        curmax = 1
        tmp = 1
        res = nums[0]
        for num in nums: 
            tmp = curmax*num
            curmax = max(curmax*num,num*curmin,num)
            curmin = min(tmp,num*curmin,num)
            res = max(res,curmax)

        return res